### DISCUSSION TITLE
Losing `this` context in component

### DISCUSSION THREAD

**Discussion by marcus-at-localhost** [alpinejs/alpine#4017] [1. Help] [answered]
Hi, I'm writing an Alpine component that uses Leaflet.js for rendering a map and I'm stuck with the following:

I'm using the `L.geoJSON` method that takes in a geoJson object as described in their [docs](https://leafletjs.com/examples/geojson/)

I moved the `onEachFeature` method out of `L.geoJSON` into the Alpine component scope to reuse it, but then I'm losing the context to `this` to call `this.data`. 

Here is a stripped down example and a link to a fiddle below

```js
document.addEventListener("alpine:init", () => {
	Alpine.data('switchboard', () => ({
        data: null,
		init() {
			this.pois = L.geoJSON(this.geojson, {
				onEachFeature: this.onEachFeature,
			}).addTo(this.map);
		},
		onEachFeature: (feature, layer) => {
			// for some reason we got no access to this.data from the component,
			// most likely because it's called within a leaflet method.
			// this refers to the window object
			console.log(this.data);
		}
	}))
});
```
https://jsfiddle.net/u5zxp84k/26/

I tried some variations of `.bind(this)` and even hoped that I could use something like `Alpine.$data.data`, but I'm as lost as `this` in this context ^^.

Thanks for any hints.

**Comment by marcus-at-localhost** [comment]
I did it this way for the time being

```js
document.addEventListener("alpine:init", () => {
	Alpine.data('switchboard', () => ({
        data: null,
		init() {
			this.pois = L.geoJSON(this.geojson, {
				onEachFeature: (feature, layer) => {
					this.onEachFeature(feature, layer, this.data);
				},
			}).addTo(this.map);
		},
		onEachFeature: (feature, layer, data) => {
			console.log(data);
		}
	}))
});
```

But I'd like to understand how to keep `this` in this context. Would be Stack Overflow a better place to ask?

**Comment by ekwoka** [comment]
Easy, don't have `onEachFeature` be an arrow function. Make it a normal method

```js
onEachFunction(feature, layer) {
  this.data
}
```

**Reply by marcus-at-localhost** [reply]
Now `this` looks like the `L.geoJSON` method.  https://jsfiddle.net/nzuv0aLt/5/

![image](https://github.com/alpinejs/alpine/assets/1337223/6c153241-e8ba-475c-b79e-45faa8961ce5)

**Comment by marcus-at-localhost** [comment, marked as answer]
Ok, I think I got the right combination of how to write the `onEachFeature` method and `bind(this)` 

```js
document.addEventListener("alpine:init", () => {
	Alpine.data('switchboard', () => ({
        data: null,
		init() {
			this.pois = L.geoJSON(this.geojson, {
				onEachFeature: this.onEachFeature.bind(this), // <- bind(this)
			}).addTo(this.map);
		},
		onEachFeature: function(feature, layer) { // <- normal function/method
			console.log(this.data);
		}
	}))
});
```

https://jsfiddle.net/7n5g0yux/2/

I think I got some learning to do why, when and how to use arrow function (the pass-through of `this` is the most obvious one, but didn't work in this case).

Thanks @ekwoka for pointing in the right direction.

**Reply by ekwoka** [reply]
You don't need to use the function keyword.

You can just define it as a method like you did `init`. That's more typical.
